GX-TimeFrequency/src/minimax_grids.F90

```diff
@@ -0,0 +1,465 @@
+    integer, parameter                                :: cos_w_to_cos_t = 2
+    integer, parameter                                :: sin_t_to_sin_w = 3
+    integer                                           :: i_point, j_point
+    real(kind=dp)                                     :: e_range, scaling
+    real(kind=dp), dimension(:), allocatable          :: x_tw
+    real(kind=dp), dimension(:, :), allocatable       :: mat
+
+    ! Begin work
+    e_range = e_max/e_min
+    ierr = 0
+
+    allocate (x_tw(2*num_points))
+
+    call get_points_weights_omega(num_points, e_range, x_tw, ierr)
+    if (ierr /= 0) return
+
+    allocate (omega_points(num_points))
+    allocate (omega_weights(num_points))
+
+    do i_point = 1, num_points
+       omega_points(i_point) = x_tw(i_point)
+       omega_weights(i_point) = x_tw(i_point + num_points)
+    end do
+
+    ! scale the minimax parameters
+    omega_points(:) = omega_points(:)*e_min
+    omega_weights(:) = omega_weights(:)*e_min
+
+    ! set up the minimax time grid
+    call get_points_weights_tau(num_points, e_range, x_tw, ierr)
+    if (ierr /= 0) return
+
+    ! For RPA we include already a factor of two (see later steps)
+    scaling = 2.0_dp
+
+    allocate (tau_points(num_points))
+    allocate (tau_weights(num_points))
+
+    do i_point = 1, num_points
+       tau_points(i_point) = x_tw(i_point)/scaling
+       tau_weights(i_point) = x_tw(i_point + num_points)/scaling
+    end do
+
+    ! scale grid from [1,R] to [e_min,e_max]
+    tau_points(:) = tau_points(:)/e_min
+    tau_weights(:) = tau_weights(:)/e_min
+
+    allocate (cosft_wt(num_points, num_points))
+    allocate (cosft_tw(num_points, num_points))
+    allocate (sinft_wt(num_points, num_points))
+
+    ! get the weights for the cosine transform W^c(it) -> W^c(iw)
+    call get_transformation_weights(num_points, tau_points, omega_points, cosft_wt, e_min, e_max, &
+         max_errors(1), cos_t_to_cos_w, ierr)
+    if (ierr /= 0) return
+
+    ! get the weights for the cosine transform W^c(iw) -> W^c(it)
+    call get_transformation_weights(num_points, tau_points, omega_points, cosft_tw, e_min, e_max, &
+         max_errors(2), cos_w_to_cos_t, ierr)
+    if (ierr /= 0) return
+
+    ! get the weights for the sine transform Sigma^sin(it) -> Sigma^sin(iw) (PRB 94, 165109 (2016), Eq. 71)
+    call get_transformation_weights(num_points, tau_points, omega_points, sinft_wt, e_min, e_max, &
+         max_errors(3), sin_t_to_sin_w, ierr)
+    if (ierr /= 0) return
+
+    ! Compute the actual weights used for the inhomogeneous cosine/ FT and check whether
+    ! the two matrices for the forward/backward transform are the inverse of each other.
+    do i_point = 1, num_points
+       do j_point = 1, num_points
+          cosft_wt(j_point, i_point) = cosft_wt(j_point, i_point)*cos(tau_points(i_point)*omega_points(j_point))
+          cosft_tw(i_point, j_point) = cosft_tw(i_point, j_point)*cos(tau_points(i_point)*omega_points(j_point))
```

Comment:
I would build this guy in a separately as the memory access is suboptimal. Maybe not a big deal though:

```
do j_point = 1, num_points
  do i_point = 1, num_points
     cosft_tw(i, j)
```

Reply:
We can keep  as it is